### DISCUSSION TITLE
Future of OSx: Apple vs Customers and Developers?

### DISCUSSION THREAD

**Discussion by WarthogARJ** [Homebrew/discussions#3589] [Everyday usage]
Hi all,
This is a somewhat philosophical question/comment.

If you look at the development of Apple OSx, the 3rd party software for it, and the related hardware to run it all, it's gone through a pretty steady change ("de-evolution"?) over the past 20-30 years that's made things less flexible, and allowed the Customer and 3rd Party Developers less lattitude:
Hardware: Power PC to Intel to ARM
Architecture: 32 bit only, to both 32/64 bit allowed, to only 64 bit
Ability to run non-OSx on hardware: it's gotten more restrictive: at the moment it is not easy to boot into Windows, or even run a VM with Windows on it
Graphics: it's become quite restrictive about what GPU's you install, assuming you can swap them in. And Apple has made it close to impossible to run anything but Metal on the newer machines/OSx

In general, if you want to be able to continue running either your existing software, or to add new ones, you tend to be forced to "upgrade" your OSx so that it's no more than 4-5 years old (for example, you will struggle if you try to use anything < Catalina now).

If you buy new hardware, it's all ARM-based, and most (almost all) of the systerms do not allow you to replace or repair any component, let alone upgrade it to something not OEM. 
As in you now have very little leeway in changing CPU, RAM, SSD or GPU.
If a component fails, if you are not covered by AppleCare, you are looking at a lot of trouble in most cases.

Is this a problem?
Well, to ME it seems to be one.
Apple hardware is pretty good, and you can use your MacBook/MacMini/iMac etc for a LONG time.
5 years easily, and in many cases a lot longer.
And that's well known, so there is a strong 2nd hand market.

But this is going to change in the next 3-4 years.
At some point a lot of the Intel-based Apple hardware will not run the newer OSx, and we'll be forced to go to ARM.
And a lot of the software that ran happily on Intel will need major re-writes to run on the newer OSx/ARM systems.

It used to be that buying an Apple meant you had a good pairing of hardware and OS, and yet you could STILL run other stuff on it: Windows, Linux etc etc.
And....it WORKED.

But now?
well, it's become more a case of "it doesn't work".
A few examples:
- nVidia was forced out, as in you cannot run CUDA.
- It's difficult to run OpenCL
- cannot run BootCamp
- VM's that use x86-64 platforms

It seems to me that this general mode of behaviour by Apple is anethema to ethos underlaying groups like HomeBrew and open source in general.
It doesn't make things impossible overall, just more and more difficult: for both Customers and 3rd Party Developers.
I think we lost more than we gained when Apple went this way.
In the short-term, Apple gains.
But the rest of us lost: both short-term AND long-term; unless we complain and get Apple to be more open-minded.

Perhaps HomeBrew should be permantently ported to Linux, and the Apple ties deprecated....:-}

**Comment by p-linnane** [comment]
I don't really understand the purpose of this post being in the Homebrew discussions, but sure. 

>In general, if you want to be able to continue running either your existing software, or to add new ones, you tend to be forced to "upgrade" your OSx so that it's no more than 4-5 years old (for example, you will struggle if you try to use anything < Catalina now).

Yes, this is how things work in the modern world. 4-5 years is an eternity when it comes to the speed at which tech moves. The security implications of running a 5 year old unsupported OS alone are mortifying. Or at least they should be to anyone who has a basic grasp of cybersecurity in 2022.

>If you buy new hardware, it's all ARM-based, and most (almost all) of the systerms do not allow you to replace or repair any component, let alone upgrade it to something not OEM.
As in you now have very little leeway in changing CPU, RAM, SSD or GPU.
If a component fails, if you are not covered by AppleCare, you are looking at a lot of trouble in most cases.

This has nothing to do with Homebrew. If you don't like the hardware Apple provides, don't use it.

>But this is going to change in the next 3-4 years.
At some point a lot of the Intel-based Apple hardware will not run the newer OSx, and we'll be forced to go to ARM.
And a lot of the software that ran happily on Intel will need major re-writes to run on the newer OSx/ARM systems.

We do not have a timeline for deprecation of macOS running on Intel. Rosetta 2 is making things easier right now and many applications have released arm64 builds either as stand alone from the x86_64 or as Universal binaries. It is disingenuous to make this kind of claim in such a broad sense.

>It seems to me that this general mode of behaviour by Apple is anethema to ethos underlaying groups like HomeBrew and open source in general.
It doesn't make things impossible overall, just more and more difficult: for both Customers and 3rd Party Developers.
I think we lost more than we gained when Apple went this way.
In the short-term, Apple gains.
But the rest of us lost: both short-term AND long-term; unless we complain and get Apple to be more open-minded.

This has nothing to do with Homebrew. Apple is a public corporation. They do what they do for profit. Their software is generally closed source with open source components. It's strange to make this point as if Apple has always released their source code and is now changing that. 

>Perhaps HomeBrew should be permantently ported to Linux, and the Apple ties deprecated....:-}

Please add a link to your pull request suggesting and implementing this. I'd love to see that review process.

**Comment by SMillerDev** [comment]
>  And a lot of the software that ran happily on Intel will need major re-writes to run on the newer OSx/ARM systems.

Almost all software that is maintained is already available on ARM. Apple is not the only company moving towards ARM, the whole IT world is. ARM has been a huge part of the software world since ehh... the original iPhone 15 years ago.

>  It doesn't make things impossible overall, just more and more difficult: for both Customers and 3rd Party Developers.

As a maintainer, I've not noticed things becoming more difficult.

>  I think we lost more than we gained when Apple went this way.

And you'd be wrong, all software builds much faster on ARM so from a Homebrew perspective this is better.

> Perhaps HomeBrew should be permantently ported to Linux

It's been official on Linux since 2019: https://brew.sh/2019/02/02/homebrew-2.0.0/

> and the Apple ties deprecated

So you're suggesting removing a useful tool from the Apple ecosystem because you disagree with the repairability of Apple products?  That "doesn't make things impossible overall, just more and more difficult: for both Customers and 3rd Party Developers." so it would be the exact thing you're accusing Apple of.

**Reply by WarthogARJ** [reply]
Good point: I'm being a bit illogical and inconsisent here.

My main bug bear is that I've used a lot of niche software that's a lot harder to run in the newer environments.
But perhaps I should just set-up a series of VM's with a fixed set of OS/XCode, Homebrew etc that is consistent. 
And then not update/upgrade it.

And will run whatever software that hasn't been upgraded, where the developer decided to allow it to go obsolete.

As far as hardware goes, I'm not anti-ARM, but I don't like the general trend to making it hard to change the system components after purchase: either as an upgrade, or to repair it.
If you look at the cost of a new Mac laptop, it's a VERY big cost relative to the "average" person's disposable income.
And I mean a laptop with more than the very small standard OEM RAM and SSD. Let alone going to the higher capablity processor.

You used to be able to buy a used Mac, and upgrade it to your own preferences: but no longer.
You cannot even repair it very easily as a 2nd owner.
And that's been done on purpose: as a marketing exercise.
There's no technical reason to do it.

**Reply by SMillerDev** [reply]
> If you look at the cost of a new Mac laptop, it's a VERY big cost relative to the "average" person's disposable income.

MacBooks have always been expensive.

> You used to be able to buy a used Mac, and upgrade it to your own preferences: but no longer.

Maybe I'm too young to know this, but I've never heard of any laptop where this was supported

**Comment by edrozenberg** [comment]
HomeBrew started on Mac and I hope it will continue to help us Mac users for many more years to come. Nice to see Linux Brew of course, but Linux already has so many packaging systems (at least dozens), and each distro its own packaging preferences and conventions. HomeBrew is much more important to the Mac than SomeOtherBrew is or likely will be, to any other platforms.

Most Mac apps already provide ARM builds - on an ARM Mac you can use iMazing's free **Silicon** app to see which ones don't yet https://github.com/DigiDNA/Silicon.

VMWare Fusion is behind and very slow in fully supporting ARM Macs (can't run MacOS yet for ex). But you can use the free UTM (v4 in beta, some issues atm - https://github.com/utmapp/UTM/releases) to run MacOS on ARM Macs with fantastic performance, since it uses the new Apple VM functionality. Running MacOS VM's this way is a much much better performing solution than MacOS VM's on VMWare Fusion Intel ever was.

The main reason Windows VM's on ARM Macs isn't happening is an MS licensing issue - they're not anxious to make this possible. Believe there may be some workarounds available.

For Macs "left behind", the OpenCore Patcher brings Big Sur/Monterrey support to many older Macs. Something I'm planning to do on an old iMac (req wiping the machine, so prep work...). https://github.com/dortania/OpenCore-Legacy-Patcher . Not sure about Ventura support yet, maybe (likely?) will come.

The Mac App store hasn't made a big splash, and so the fears of software lockdown on the Mac have not come true. There are more restrictions now than there were in the past (signing etc), but you can still run any software you want, and hopefully the remaining stragglers will get around to providing native ARM builds soon.

Stuff that's a bit worrying is the Swift UI redesign of some core things like the Prefs panel and print panel in Ventura, which many are complaining about. I'd be happier if they didn't mess with the Mac too much, iOSification isn't what many of us want, but hopefully it won't go too far.

**Reply by WarthogARJ** [reply]
Thanks very much for your replies.
Perhaps I'm just expecting too much from a given mix of hardware/OS.

As for why post this in Homebrew, it's because I see the Homebrew environment as one of reasons for using an Apple system.

And thanks for the info on UTM, and OPenCore Patcher.
My general understanding is obviously hindered by not owning an Apple ARM system: but the sticker shock for a new M1/M2 laptop for me is just too big.
